File: Languages/cpp/algorithms-dsa/leetcode/dynamic-programming/house_robber.cpp
/*
 * LeetCode 198 - House Robber
 * Topic: Dynamic Programming
 * Difficulty: Medium
 *
 * At each house decide to rob (prev2 + current) or skip (prev1). O(1) space DP.
 * Time: O(n)
 * Space: O(1)
 */
#include <iostream>
#include <vector>
#include <cassert>
using namespace std;

int rob(vector<int>& nums) {
    int prev2 = 0, prev1 = 0;
    for (int n : nums) {
        int curr = max(prev1, prev2 + n);
        prev2 = prev1;
        prev1 = curr;
    }
    return prev1;
}

int main() {
    vector<int> t1 = {1,2,3,1};
    assert(rob(t1) == 4);

    vector<int> t2 = {2,7,9,3,1};
    assert(rob(t2) == 12);

    vector<int> t3 = {2,1,1,2};
    assert(rob(t3) == 4);

    vector<int> t4 = {0};
    assert(rob(t4) == 0);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/dynamic-programming/house_robber.py
"""
LeetCode #198 - House Robber
Topic: Dynamic Programming
Difficulty: Medium

Maximize robbery amount where adjacent houses cannot both be robbed (take or skip).

Time Complexity: O(n)
Space Complexity: O(1)
"""


def rob(nums: list[int]) -> int:
    prev, curr = 0, 0
    for num in nums:
        prev, curr = curr, max(curr, prev + num)
    return curr


if __name__ == "__main__":
    assert rob([1, 2, 3, 1]) == 4
    assert rob([2, 7, 9, 3, 1]) == 12
    assert rob([2, 1, 1, 2]) == 4
    assert rob([0]) == 0
    print("All tests passed!")
